Index: mediawiki_filter.settings.inc
===================================================================
--- mediawiki_filter.settings.inc	(revision 2653)
+++ mediawiki_filter.settings.inc	(working copy)
@@ -206,7 +206,7 @@
     if ($line) {
       list($title, $id, $url) = explode('|', $line, 3);
       if (trim($url) == '<path>') {
-        $url = '$1'; //str_replace('__QUERY_POSITION__', '$1', url('__QUERY_POSITION__', NULL, NULL, TRUE));
+        $url = '$1'; //str_replace('__QUERY_POSITION__', '$1', url('__QUERY_POSITION__', array ('absolute' => TRUE)));
       }
       if ($op == 'url') {
         $mappings[trim($id)] = trim($url);
Index: mediawiki_filter.info
===================================================================
--- mediawiki_filter.info	(revision 2653)
+++ mediawiki_filter.info	(working copy)
@@ -1,3 +1,7 @@
 ; $Id$
 name = "Mediawiki filter"
 description = "Filter which uses Mediawiki parser for formatting"
+package = Filters
+version = "6.x-0.1"
+core = "6.x"
+
Index: mediawiki_filter.lib.inc
===================================================================
--- mediawiki_filter.lib.inc	(revision 2653)
+++ mediawiki_filter.lib.inc	(working copy)
@@ -63,7 +63,7 @@
 
   // Set up some global variables needed for Mediawiki.
   $settings = array(
-    'wgServer' => url(NULL, NULL, NULL, TRUE),
+    'wgServer' => url(NULL, array('absolute' => TRUE)),
     'wgScriptPath' => '/xyz',
     'wgScript' => '/xyz/index.php',
     'wgRedirectScript' => '/xyz/redirect.php',
@@ -696,7 +696,9 @@
         $url .= "?$query";
       }
     } else {
+      $raw_url = $this->getPrefixedDBkey();
       $dbkey = wfUrlencode( $this->getPrefixedDBkey() );
+
       if ( $query == '' ) {
         if($variant!=FALSE && $wgContLang->hasVariants()){
           if($wgVariantArticlePath==FALSE) {
@@ -708,7 +710,12 @@
           $url = str_replace( '$1', $dbkey, $url  );
         }
         else {
-          $url = str_replace( '$1', $dbkey, $wgArticlePath );
+          /* FIXME: JFIAT */
+          $l_url = $raw_url;
+          if (module_exists('pathauto')) {
+	    $l_url = pathauto_cleanstring($l_url);
+          }
+          $url = str_replace( '$1', $l_url, $wgArticlePath );
         }
       } else {
         global $wgActionPaths;
@@ -732,7 +739,7 @@
           $url = "{$wgScript}?title={$dbkey}&{$query}";
         }
       }
-      
+
       // FIXME: this causes breakage in various places when we
       // actually expected a local URL and end up with dupe prefixes.
 // TODO: For Drupal we always render, no matter what
@@ -1566,7 +1573,7 @@
       $image_sizes = array();
     }
     if (isset($params['width'])) {
-      foreach($image_sizes as $image_size) {
+      foreach($image_sizes as $image_label => $image_size) {
         $image_label = $image_size['label'];
         if ($image_size['width'] >= $params['width']) {
           break;
@@ -1764,4 +1771,3 @@
     return NULL;
   }
 }
-
Index: mediawiki_filter.module
===================================================================
--- mediawiki_filter.module	(revision 2653)
+++ mediawiki_filter.module	(working copy)
@@ -12,8 +12,7 @@
 /**
  * Implementation of hook_menu().
  */
-function mediawiki_filter_menu($may_cache) {
-  if (!$may_cache) {
+function mediawiki_filter_menu() {
     drupal_add_css(drupal_get_path('module', 'mediawiki_filter') .'/mediawiki_filter.css');
 
     // TODO: make external image an option
@@ -28,7 +27,6 @@
         });
       }',
       'inline', 'footer');
-  }
 }
 
 /**
@@ -218,7 +216,7 @@
         $url = url(str_replace('$1', $title, $url));
         break;
       case 'create-link':
-        $url = url('node/add/' . $type, 'edit[title]=' . urlencode($title));
+        $url = url('node/add/' . $type, array('query' => 'edit[title]=' . urlencode($title)));
         break;
     }
   }
@@ -230,7 +228,7 @@
   $interwiki_links = mediawiki_filter_interwiki_mappings();
   if (isset($interwiki_links[$key])) {
     $url = str_replace('$1', '__QUERY_POSITION_REPLACEMENT__', $interwiki_links[$key]);
-    $url = url($url, NULL, NULL, TRUE);
+    $url = url($url, array('absolute' => TRUE));
     $url = str_replace('__QUERY_POSITION_REPLACEMENT__', '$1', $url);
     return $url;
   }
